opm/simulators/linalg/FlexibleSolver.hpp

```diff
@@ -46,16 +46,27 @@ class FlexibleSolver : public Dune::InverseOperator<VectorTypeT, VectorTypeT>
     using VectorType = VectorTypeT;
 
     /// Create a sequential solver.
-    FlexibleSolver(const boost::property_tree::ptree& prm, const MatrixType& matrix)
+    FlexibleSolver(const boost::property_tree::ptree& prm, const MatrixType& matrix,
+                   const std::function<VectorTypeT()>& weightsCalculator = std::function<VectorTypeT()>())
     {
-        init(prm, matrix, Dune::Amg::SequentialInformation());
+        init(prm, matrix, weightsCalculator, Dune::Amg::SequentialInformation());
     }
 
     /// Create a parallel solver (if Comm is e.g. OwnerOverlapCommunication).
     template <class Comm>
-    FlexibleSolver(const boost::property_tree::ptree& prm, const MatrixType& matrix, const Comm& comm)
+    FlexibleSolver(const boost::property_tree::ptree& prm,
+                   const typename std::enable_if<std::is_function<Comm>::value,MatrixType>::type& matrix,
```

Comment:
I think this is too complicated (I did understand it after a while). I realize that we need to distinguish the third argument being a weight calculator vs. it being a communication object. But this is wrong is it not? On line 61, you should pass comm as the third argument here? What about putting the Comm argument first, for the parallel constructor? Then there would be no ambiguity and no need for an extra constructor.

I think encapsulating the optional weights calculator in a std::function object is a good improvement (it is a bit of a pain in any case).

Reply:
On line 61 `comm` is the fourth argument. Compiler that yes and seems to be right most of the time (which I sometimes regret). 
The thought of putting comm first, did not occur to be. Probably, because I am expecting optional arguments last. And it would result in changes at many places.

Would
```

template<class C>
struct IsComm : std::false_type
{};

template<>
struct IsComm<Dune::Amg::SequentialInformation> : std::true_type
{};

template<class Index>
struct IsComm<Dune::OwnerOverlapCopyCommunication<Index>> : std::true_type
{};
template <class Comm>
FlexibleSolver(const boost::property_tree::ptree& prm,
                       const MatrixType& matrix,
                       const typename std::enable_if<IsComm<Comm>::value, Comm>::type& comm)
```
be better? I will add that.